HCK06_2019_Janelia/projects/ExtensionSharing/README.md

```diff
@@ -18,28 +18,36 @@ Develop methods for sharing NWB:N extensions.
 
 ## Objectives
 
-- Evaluate plans for infratructure for web-based sharing, deployment, and evaluation of extensions
-- Begin implmetation of the online archive for NWB:N extensions
+- Evaluate plans for infrastructure for web-based sharing, deployment, and evaluation of extensions
+- Begin implementation of the online archive for NWB:N extensions
 
 ## Approach and Plan
 
 - Work with attendees to evaluate the current plan
 - Implement draft template repo for extensions
-- Implement draft template repo for extension recipies 
+- Implement draft template repo for extension recipes
 
 ## Progress and Next Steps
 
-- Created GitHub organization to develop the NWB:N extensions catalog 
+- Created GitHub organization to develop the NWB:N extensions catalog
 - Created draft GitHub webpage for the extension catalog
 - Created design documents for the extension catalog
-- Created draft extension and sharing and versioning guidelines
-- Added continuous integration (using Azure-pipelines) to [NeurodataWithoutBorders/nwb-docutils](https://github.com/NeurodataWithoutBorders/nwb-docutils), [nwb-extensions/ndx-template](https://github.com/nwb-extensions/ndx-template)
+- Created draft extension sharing and versioning guidelines
+- Added continuous integration (using Azure-pipelines) to [NeurodataWithoutBorders/nwb-docutils](https://github.com/NeurodataWithoutBorders/nwb-docutils), [nwb-extensions/ndx-template](https://github.com/nwb-extensions/ndx-template) and  [NeurodataWithoutBorders/pynwb](https://github.com/NeurodataWithoutBorders/pynwb)
 - Consolidated documentation of `ndx-template` with the one generated by `nwb-docutils`
 - Reviewed conda-forge infrastructure and identified next step for enabling continuous-integration of [nwb-extensions/staged-extensions](https://github.com/nwb-extensions/staged-extensions)
+- Discussed the flowchart for users creating and publishing their own extensions
+  - Decided to simplify the initial creation of an extension using a web form
+  - Decided that `version` needs to be required on YAML spec files, and now that MatNWB cache spec bug has been fixed, `cache_spec` should default to True when writing an NWB file.
+  - Decided it would be nice to have a utility function for downloading an extension, e.g. `pynwb-install-ext <extension_name>` would automatically look up the source for the extension on our catalog and download the YAML files and any associated source code from PyPi or GitHub into the local filesystem.
```

Comment:
I think `pip install name-of-extension` will do that and will remove the need to implement any custom cli.

Reply:
For published Python packages, that would work, but we need to keep support for extensions not published on PyPi, e.g. on GitHub or elsewhere. The custom CLI would probably do:
- look up extension on our catalog
- process ndx-meta.yaml
- if `pip` URL specified, run `pip install name-of-extension`
- else, download and unzip source from `url` field

Reply:
Yeah I agree with Ryan that we want to make sure our pipeline works for matlab users as well who would be familiar with GitHub but not pypi

Reply:
I concur, we need a way to install extensions directly from GitHub. In particular in the case where folks don't supply custom classes, we should ideally not require folks to do the extra step of pip. And also the Matlab (and other languages) need to be able to mirror what we do in Python (e.g,. Thomas is working in Igor).

Reply:
> , but we need to keep support for extensions not published on PyPi, e.g. on GitHub or elsewhere
> we need a way to install extensions directly from GitHub

If extension source code is directly provided, or if extension is available on GitHub but not on PyPi, they could be installed using any of these:

```
pip install /path/to/source
```

or 

```
pip install git+https://github.com/org/project
```

More details at https://pip.pypa.io/en/stable/reference/pip_install/#vcs-support


> And also the Matlab (and other languages) need to be able to mirror what we do in Python

To support other languages, I highly recommend to standardize on the distribution mechanism associated with that language (gem for ruby, cargo for rust, ... )

After talking to @nclack  last week, he mentioned that to work with matnwb, having the URL to download the yaml files should be sufficient. 

That said, I wonder if this matlab package manager is used: http://mobeets.github.io/mpm/

Reply:
Huh. I didn't know that pip install had that functionality. That looks good for covering Python sources not published to PyPi.

Re: matlab: That package manager looks useful, but I have never seen it before. And now that I think of it, if I had to guess, the average Matlab user in neuroscience has little experience with the command line. For matnwb, I would write a specific script in Matlab, which when run in Matlab, would perform the steps that I outlined above, minus the `pip install` part. Matlab has some web service utilities:  https://www.mathworks.com/help/matlab/ref/websave.html . Then the user does not have to leave the Matlab application to download the yaml and source files.